File: android/src/main/java/com/dynamsoft/capacitor/dlr/LabelRecognizerPlugin.java
package com.dynamsoft.capacitor.dlr;

import android.util.Log;

import com.dynamsoft.dlr.DLRResult;
import com.dynamsoft.dlr.LabelRecognizerException;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONArray;
import org.json.JSONException;

import java.lang.reflect.InvocationTargetException;

@CapacitorPlugin(name = "LabelRecognizer")
public class LabelRecognizerPlugin extends Plugin {

    private LabelRecognizer implementation = new LabelRecognizer();

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");
        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod
    public void initLicense(PluginCall call) {
        String license = call.getString("license");
        implementation.initLicense(license,getContext());
        call.resolve();
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        try {
            implementation.initDLR();
            call.resolve();
        } catch (LabelRecognizerException e) {
            e.printStackTrace();
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void recognizeBase64String(PluginCall call) {
        String base64 = call.getString("base64");
        base64 = base64.replaceFirst("data:.*?;base64,","");
        Log.d("DLR",base64);
        try {
            JSArray results = implementation.recognizeBase64String(base64);
            JSObject response = new JSObject();
            response.put("results",results);
            call.resolve(response);
        } catch (LabelRecognizerException e) {
            e.printStackTrace();
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void recognizeBitmap(PluginCall call) {
        try {
            String className = call.getString("className","com.tonyxlh.capacitor.camera.CameraPreviewPlugin");
            String methodName = call.getString("methodName","getBitmap");
            JSArray results = implementation.recognizeBitmap(className,methodName);
            JSObject response = new JSObject();
            response.put("results",results);
            call.resolve(response);
        } catch (LabelRecognizerException | InvocationTargetException | IllegalAccessException |
                 NoSuchMethodException | ClassNotFoundException e) {
            e.printStackTrace();
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void updateRuntimeSettings(PluginCall call) {
        JSObject settings = call.getObject("settings");
        String template = settings.getString("template");
        try {
            implementation.updateRuntimeSettings(template);
        } catch (LabelRecognizerException e) {
            e.printStackTrace();
        }
        if (settings.has("customModelConfig")) {
            JSObject config = settings.getJSObject("customModelConfig");
            String modelFolder = config.getString("customModelFolder");
            try {
                JSONArray fileNames = config.getJSONArray("customModelFileNames");
                implementation.loadCustomModel(getContext(), modelFolder, fileNames);
            } catch (JSONException | LabelRecognizerException e) {
                e.printStackTrace();
            }
        }
        call.resolve();
    }

    @PluginMethod
    public void resetRuntimeSettings(PluginCall call) {
        try {
            implementation.resetRuntimeSettings();
            call.resolve();
        } catch (LabelRecognizerException e) {
            e.printStackTrace();
            call.reject(e.getMessage());
        }
    }
}

File: ios/Plugin/LabelRecognizerPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(LabelRecognizerPlugin)
public class LabelRecognizerPlugin: CAPPlugin  {
    private let implementation = LabelRecognizer()

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }
    
    @objc func initialize(_ call: CAPPluginCall) {
        implementation.initialize()
        call.resolve()
    }
    
    @objc func initLicense(_ call: CAPPluginCall) {
        let license = call.getString("license") ?? "DLS2eyJoYW5kc2hha2VDb2RlIjoiMjAwMDAxLTE2NDk4Mjk3OTI2MzUiLCJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSIsInNlc3Npb25QYXNzd29yZCI6IndTcGR6Vm05WDJrcEQ5YUoifQ=="
        implementation.initLicense(license)
        call.resolve()
    }
    
    @objc func recognizeBase64String(_ call: CAPPluginCall) {
        var base64 = call.getString("base64") ?? ""
        base64 = removeDataURLHead(base64)
        call.resolve(["results":implementation.recognizeBase64String(base64)])
    }
    
    func removeDataURLHead(_ str: String) -> String {
        var finalStr = str
        do {
            let pattern = "data:.*?;base64,"
            let regex = try NSRegularExpression(pattern: pattern, options: NSRegularExpression.Options.caseInsensitive)
            finalStr = regex.stringByReplacingMatches(in: str, options: NSRegularExpression.MatchingOptions(rawValue: 0), range: NSMakeRange(0, str.count), withTemplate: "")
        }
        catch {
            print(error)
        }
        return finalStr
    }
    
    @objc func updateRuntimeSettings(_ call: CAPPluginCall) {
        let settings = call.getAny("settings") as? [String:Any]
        let template = settings!["template"] as! String
        implementation.updateRuntimeSettings(template)
        
        guard let customModelConfig = settings!["customModelConfig"] as? [String:Any] else {
            call.resolve()
            return
        }
        let modelFolder = customModelConfig["customModelFolder"] as! String
        let modelFileNames = customModelConfig["customModelFileNames"] as! [String]
        implementation.loadCustomModel(modelFolder: modelFolder, modelFileNames: modelFileNames)
        call.resolve()
    }
    
    @objc func resetRuntimeSettings(_ call: CAPPluginCall) {
        implementation.resetRuntimeSettings()
        call.resolve()
    }
}
